src/general_tally.cpp

```diff
@@ -0,0 +1,210 @@
+#include <tallies/general_tally.hpp>
+#include <utils/error.hpp>
+
+GeneralTally::GeneralTally(std::shared_ptr<PositionFilter> position_filter,
+                           std::shared_ptr<EnergyFilter> energy_in,
+                           Quantity quantity, Estimator estimator,
+                           std::string name_)
+    : ITally(quantity, estimator, name_),
+      position_filter_(position_filter),
+      energy_in_(energy_in) {
+  // flag variable to predict weather position and energy filter exists or not
+  bool check_ifany_filter = true;
+
+  // get the shape or dimension of position filter
+  StaticVector3 position_shape_;
+  if (position_filter_) {
+    position_shape_ = position_filter_->get_shape();
+    check_ifany_filter = false;
+  }
+
+  StaticVector4 tally_dimensions_(position_shape_.begin(),
+                                  position_shape_.end());
+
+  // get the size of the enery_filter
+  if (energy_in_) {
+    size_t ne = energy_in_->size();
+    tally_dimensions_.insert(tally_dimensions_.begin(), ne);
+
+    check_ifany_filter = check_ifany_filter && false;
+  }
+
+  // if any of the filter is not present, then give fatal_error
+  if (check_ifany_filter == true) {
+    fatal_error("for the " + name_ + " tally, no filter is provided.");
+  }
+
+  tally_avg.reallocate(tally_dimensions_);
+  tally_avg.fill(0.0);
+
+  tally_gen_score.reallocate(tally_dimensions_);
+  tally_gen_score.fill(0.0);
+
+  tally_var.reallocate(tally_dimensions_);
+  tally_var.fill(0.0);
+}
+
+void GeneralTally::score_collision(const Particle& p, const Tracker& tktr,
+                                   MaterialHelper& mat) {
+  if (!(estimator_ == GeneralTally::Estimator::Collision)) {
+    return;
+  }
+
+  // get the indexes for the positions, if exist
+  StaticVector3 position_indexes;
+  if (position_filter_) {
+    position_indexes = position_filter_->get_indices(
+        tktr);  // it will provde the reduce dimensions
+  }
+  StaticVector4 indexes(position_indexes.begin(), position_indexes.end());
+
+  // get the index for the energy if exist
+  std::size_t index_E;
+  if (energy_in_) {
+    std::optional<std::size_t> E_indx = energy_in_->get_index(p.E());
+    if (E_indx.has_value()) {
+      index_E = E_indx.value();
+    } else {
+      return;
+    }
+  }
+
```

Comment:
I think we need a check here, where if `position_filter_ == nullptr && energy_in_ == nullptr`, we need to set `indexes = {0};`, as we should integrate over all position and energy.